cirq-core/cirq/transformers/dynamical_decoupling.py

```diff
@@ -0,0 +1,146 @@
+# Copyright 2024 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Transformer pass that adds dynamical decoupling moments to a circuit."""
+
+import enum
+from functools import reduce
+from typing import Any, Dict, Optional, Tuple
+
+import cirq
+from cirq import value
+import numpy as np
+
+
+@enum.unique
+class _DynamicalDecouplingSchema(enum.Enum):
+    """Supported schemes of dynamical decoupling."""
+
+    XX_PAIR = 'XX_PAIR'
+    YY_PAIR = 'YY_PAIR'
+
+
+def _repeat_sequence(base_sequence: list['cirq.Gate'], num_idle_moments: int):
+    repeat_times = num_idle_moments // len(base_sequence)
+    return base_sequence * repeat_times
+
+
+def _generate_dd_sequence_from_schema(
+    schema: _DynamicalDecouplingSchema, num_idle_moments: int = 2
+) -> list['cirq.Gate']:
+    match schema:
+        case _DynamicalDecouplingSchema.XX_PAIR:
+            return _repeat_sequence([cirq.XPowGate(), cirq.XPowGate()], num_idle_moments)
+        case _DynamicalDecouplingSchema.YY_PAIR:
+            return _repeat_sequence([cirq.YPowGate(), cirq.YPowGate()], num_idle_moments)
+
+
+def _validate_dd_sequence(dd_sequence: list['cirq.Gate']) -> None:
+    if len(dd_sequence) < 2:
+        raise ValueError('Invalid dynamical decoupling sequence. Expect more than one gates.')
+    matrices = [cirq.unitary(gate) for gate in dd_sequence]
+    product = reduce(np.matmul, matrices)
+    if not np.array_equal(product, np.eye(2)):
```

Comment:
You could use `cirq.equal_up_to_global_phase()` here (since we don't care if the sequence differs from the identity by a global phase).

Reply:
Done.